packages/eslint-plugin-wpcalypso/lib/rules/no-package-relative-imports.js

```diff
@@ -0,0 +1,128 @@
+const fs = require( 'fs' );
+const path = require( 'path' );
+
+/**
+ * Returns the imports relative to a directory.
+ * For example, if the FS is:
+ *   - dir/
+ *      └ subdir/
+ *      └ file.js
+ *      └ data.json
+ *
+ * It will return the array ['subdir', 'file', 'data.json']
+ */
+const getRelativeImports = ( relativeDir ) => {
+	return fs.readdirSync( relativeDir, { withFileTypes: true } ).map( ( file ) => {
+		const fileName = file.name;
+		if ( file.isDirectory() ) return fileName;
+		const extension = path.extname( fileName );
+		if ( extension !== '.js' ) return fileName;
+		return path.basename( fileName, extension );
+	} );
+};
+
+/**
+ * For each import (CJS and ESM), check if the import could be relative to one of the
+ * configured `mapping` (eg: if you are importing 'X' and 'X' is a subdirectory of ./client)
+ * If it is, prepend the import with the module (eg: `import 'X'` becomes `import 'wp-client/X'`)
+ */
+const reportIfPackageRelative = ( { context, node, dependency } ) => {
+	const { mapping, warnOnDynamicImport } = context.options[ 0 ];
+
+	// Because ExportNamedDeclaration and ExportAllDeclaration may not have an imported source (aka node.source)
+	if ( ! dependency ) return;
+	const { value: dependencyName, type: dependencyType } = dependency;
+
+	// This part is only relevant for CJS imports
+	if ( dependencyType !== 'Literal' ) {
+		if ( warnOnDynamicImport ) {
+			// eslint-disable-next-line no-console
+			console.warn(
+				`Invalid import, only literal imports are supported. File ${ context.getFilename() }:${
+					node.loc.start.line
+				}`
+			);
+		}
+		return;
+	}
+
+	if ( typeof dependencyName !== 'string' ) return;
+
+	mapping.forEach( ( { dir, module } ) => {
+		const deps = getRelativeImports( dir );
+
+		if ( deps.includes( dependencyName.split( '/' )[ 0 ] ) ) {
```

Comment:
would we ant to use any kind of path resolution here to avoid rare-but-possible cases like `./././././thing/`

Reply:
Added a check of imports starting with `.` or `/` following Node's resolution algorithm. If it starts with any of those characters, then the import is ignored. If not, it doesn't really matter if it is something crazy like `foo/../../bar` (is that even valid?), it will be replaced with `wp-calypso/foo/../../bar` if `./client/foo` exists.